File: NeuralNetNative/BackPropagationAlgorithm.cpp
#define NEURALNETNATIVEAPI
#include <mathimf.h>
#include "BackPropagationAlgorithm.h"
#include "TrainPair.h"
#include "malloc.h"
#include <tbb\tbb.h>
#include <tbb\task_scheduler_init.h>
#include <tbb\parallel_for.h>
#include <tbb\blocked_range.h>
#include "GrainSizeForParallel.h"

using namespace StandardTypesNative;
using namespace tbb;

namespace NeuralNetNative {
	namespace MultyLayerPerceptron {
		BackPropagationAlgorithm::BackPropagationAlgorithm(StandardTypesNative::TrainPair **trainData, int trainDataSize) {
			_neuralNet = 0;
			_trainDataIterator = new RandomAccessIterator<TrainPair*>(trainData, trainDataSize);
			_testData = 0;
		}

		BackPropagationAlgorithm::BackPropagationAlgorithm(StandardTypesNative::TrainPair **trainData, int trainDataSize, StandardTypesNative::TrainPair **testData, int testDataSize) {
			_neuralNet = 0;
			_trainDataIterator = new RandomAccessIterator<TrainPair*>(trainData, trainDataSize);
			_testData = testData;
			_testDataSize = testDataSize;
		}

		BackPropagationAlgorithm::~BackPropagationAlgorithm(void) {
			delete _trainDataIterator;
			ClearData();
		}

		void BackPropagationAlgorithm::InitilazeMethod(NeuralNet *neuralNet, TrainProperties *trainProperties) {
			_neuralNet = dynamic_cast<MultyLayerPerceptron*>(neuralNet);
			if (_neuralNet == 0) {
				return;
			}
			_properties = trainProperties;
			_layers = _neuralNet->GetLayers();
			_layersCount = _neuralNet->GetLayersCount();
			_inputSize = _neuralNet->GetInputSize();
			_outputSize = _neuralNet->GetOutputSize();
			_packageFactor = 1.0f/_properties->PackageSize;
			_packagesCount = CalculatePackagesCount();
			AllocateMemory();
			ProcessSate = IterativeProcessState::NotStarted;
		}

		int BackPropagationAlgorithm::CalculatePackagesCount() {
			int packagesCount = _trainDataIterator->Size()/_properties->PackageSize;
			if (_trainDataIterator->Size()%_properties->PackageSize != 0) {
				packagesCount++;
			}
			return packagesCount;
		}

		TrainProperties* BackPropagationAlgorithm::Properties(void) const {
			return _properties;
		}

		void BackPropagationAlgorithm::AllocateMemory(void) {
			int maxGradientSize = FindMaxSize();

			_gradients = (float*)_mm_malloc(maxGradientSize*sizeof(float), 32);
			_gradientsIntermediate = (float*)_mm_malloc(maxGradientSize*sizeof(float), 32);

			int outputSize = _layers[_layersCount - 1]->GetSize();
			_neuronNetOutput =  (float*)_mm_malloc(outputSize*sizeof(float), 32);
			_partialDerivaitve = (float*)_mm_malloc(outputSize*sizeof(float), 32);

			_oldDeltaWeights = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_derivativeAverages = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_packageDerivative = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_learnFactors = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			for (int i = 0; i < _layersCount; i++) {
				BaseNeuralBlock *layer = _layers[i];
				int weightsCount = (layer->GetPreviousSize())*(layer->GetSize());

				_oldDeltaWeights[i] = (float*)_mm_malloc(weightsCount*sizeof(float), 32);
				_derivativeAverages[i] = (float*)_mm_malloc(weightsCount*sizeof(float), 32);
				_packageDerivative[i] = (float*)_mm_malloc(weightsCount*sizeof(float), 32);
				_learnFactors[i] = (float*)_mm_malloc(weightsCount*sizeof(float), 32);
				for (int j = 0; j < weightsCount; j++) {
					_oldDeltaWeights[i][j] = 0.0f;
					_derivativeAverages[i][j] = 0.0f;
					_packageDerivative[i][j] = 0.0f;
					_learnFactors[i][j] = 1.0f;
				}
			}

			_oldDeltaWeightsForBias = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_derivativeAveragesForBias = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_packageDerivativeForBias = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			_learnFactorsForBias = (float**)_mm_malloc(_layersCount*sizeof(float*), 32);
			for (int i = 0; i < _layersCount; i++) {
				int layerSize = _layers[i]->GetSize();

				_oldDeltaWeightsForBias[i] = (float*)_mm_malloc(layerSize*sizeof(float), 32);
				_derivativeAveragesForBias[i] = (float*)_mm_malloc(layerSize*sizeof(float), 32);
				_packageDerivativeForBias[i] = (float*)_mm_malloc(layerSize*sizeof(float), 32);
				_learnFactorsForBias[i] = (float*)_mm_malloc(layerSize*sizeof(float), 32);
				
				for (int j = 0; j < layerSize; j++) {
					_oldDeltaWeightsForBias[i][j] = 0.0f;
					_derivativeAveragesForBias[i][j] = 0.0f;
					_packageDerivativeForBias[i][j] = 0.0f;
					_learnFactorsForBias[i][j] = 1.0f;
				}
			}
		}

		int BackPropagationAlgorithm::FindMaxSize(void) {
			int maxGradientSize = 0;
			for (int i = 0; i < _layersCount; i++) {
				int curLayerSize = _layers[i]->GetSize();
				if (curLayerSize > maxGradientSize) {
					maxGradientSize = curLayerSize;
				}
			}
			return maxGradientSize;
		}

		void BackPropagationAlgorithm::RunIterativeProcess() {
			float trainErrorValue = _properties->Epsilon + 1.0f;
			float testErrorValue = std::numeric_limits<float>::quiet_NaN();
			float minTestErrorValue = FLT_MAX;
			_epochNumber = 1;
			while ((ProcessSate == IterativeProcessState::InProgress) && 
				(trainErrorValue > _properties->Epsilon) && 
				(_epochNumber <= _properties->MaxIterationCount) &&
				(isnanf(testErrorValue) || fabsf(testErrorValue - minTestErrorValue) < _properties->CvLimit)) {

				TrainEpoch();

				trainErrorValue = TestModel(_trainDataIterator->Collection(), _trainDataIterator->Size());
				testErrorValue = TestModel(_testData, _testDataSize);
				if (!isnanf(testErrorValue) && (testErrorValue < minTestErrorValue)) {
					minTestErrorValue = testErrorValue;
				}

				OnIterationCompleted(_epochNumber, trainErrorValue, testErrorValue);
				_epochNumber++;
			}
			OnIterativeProcessFinished(_epochNumber);
		}

		void BackPropagationAlgorithm::ApplyResults(void) {
			if (ProcessSate == IterativeProcessState::Finished) {
				ClearData();
			}
		}

		void BackPropagationAlgorithm::ClearData(void) {
			if (_neuralNet != 0) {
				_properties = 0;
				_neuralNet = 0;
				_layers = 0;

				for (int i = 0; i < _layersCount; i++) {
					_mm_free(_oldDeltaWeights[i]);
					_mm_free(_derivativeAverages[i]);
					_mm_free(_packageDerivative[i]);
					_mm_free(_oldDeltaWeightsForBias[i]);
					_mm_free(_derivativeAveragesForBias[i]);
					_mm_free(_packageDerivativeForBias[i]);
					_mm_free(_learnFactors[i]);
					_mm_free(_learnFactorsForBias[i]);
				}
				_mm_free(_oldDeltaWeights);
				_mm_free(_derivativeAverages);
				_mm_free(_packageDerivative);
				_mm_free(_oldDeltaWeightsForBias);
				_mm_free(_derivativeAveragesForBias);
				_mm_free(_packageDerivativeForBias);
				_mm_free(_learnFactors);
				_mm_free(_learnFactorsForBias);
				_mm_free(_gradients);
				_mm_free(_gradientsIntermediate);
				_mm_free(_neuronNetOutput);
				_mm_free(_partialDerivaitve);
			}
		}

		float BackPropagationAlgorithm::TestModel(StandardTypesNative::TrainPair **data, int dataSize) {
			if (data == 0 || dataSize == 0) {
				return std::numeric_limits<float>::quiet_NaN();
			}
			else {
				float sumError = 0.0f;
				for (int i = 0; i < dataSize; i++) {
					TrainPair *trainPair = data[i];
					_neuronNetInput = trainPair->Input();
					_neuralNet->Predict(_neuronNetInput, _neuronNetOutput);
					sumError += _properties->Metrics->Calculate(trainPair->Output(), _neuronNetOutput, _outputSize);
				}
				return sumError/dataSize;
			}
		}

		void BackPropagationAlgorithm::TrainEpoch(void) {
			_trainDataIterator->RefreshRandomAccess();
			for (int i = 0; i < _packagesCount; i++) {
				TrainPackage();
			}
		}

		void BackPropagationAlgorithm::TrainPackage(void) {
			for (int i = 0; i < _properties->PackageSize; i++) {
				TrainPair *trainPair = _trainDataIterator->Next();
				_neuronNetInput = trainPair->Input();
				_neuralNet->Predict(_neuronNetInput, _neuronNetOutput);
				_properties->Metrics->CalculatePartialDerivaitve(trainPair->Output(), _neuronNetOutput, _partialDerivaitve, _outputSize);
				CollectWeightsDelta(_partialDerivaitve);
			}
			ModifyWeightsOfNeuronNet();
		}

		void BackPropagationAlgorithm::CollectWeightsDelta(const float *errrorVector) {
			const int firstLayerNumber = 0;
			int lastHiddenLayerNumber = _layersCount - 2;

			CollectWeightsDeltaOfLayer(lastHiddenLayerNumber + 1, LocalGradientForOutputLayer, errrorVector);
			for (int layerNumber = lastHiddenLayerNumber; layerNumber >= firstLayerNumber; layerNumber--) {
				CollectWeightsDeltaOfLayer(layerNumber, LocalGradientForHiddenLayer, 0);
			}
		}

		void BackPropagationAlgorithm::CollectWeightsDeltaOfLayer(int layerNum, LocalGradient localGradientfunction, const float *partialDerivaitve) {
			BaseNeuralBlock *curLayer = _layers[layerNum];
			int prevLayerSize = curLayer->GetPreviousSize();
			int curLayerSize = curLayer->GetSize();
			int nextLayerSize = (layerNum < (_layersCount - 1)) ? _layers[layerNum + 1]->GetSize() : 0;
			float *prevLayerState = (layerNum > 0) ? _layers[layerNum - 1]->GetState() : _neuronNetInput;
			float *curGradients = _gradientsIntermediate;
			float *nextGradients = _gradients;
			float *nextLayerWeights = (layerNum < (_layersCount - 1)) ? _layers[layerNum + 1]->GetWeights() : 0;
			float *packageDerivative = _packageDerivative[layerNum];
			float *packageDerivativeForBias = _packageDerivativeForBias[layerNum];

			(*localGradientfunction)(curGradients, curLayer->GetActivationFunction(), curLayer->GetState(), partialDerivaitve, nextGradients, nextLayerWeights, curLayerSize, nextLayerSize);
			
			parallel_for( blocked_range<size_t>(0, curLayerSize, BPACollectWeightsGrainSize),
			[=](const blocked_range<size_t>& r)
			{
				for (int i = r.begin(); i < r.end(); i++) {
					float localGradient = curGradients[i];
					for (int j = 0; j < prevLayerSize; j++) {
						int weightIndex = prevLayerSize*i + j;
						packageDerivative[weightIndex] -= localGradient*prevLayerState[j];
					}
					packageDerivativeForBias[i] -= localGradient;
				}
			});
						
			_gradients = curGradients;
			_gradientsIntermediate = nextGradients;
		}

		void BackPropagationAlgorithm::ModifyWeightsOfNeuronNet(void) {
			const int firstLayerNumber = 0;
			for (int layerNum = firstLayerNumber; layerNum < _layersCount; layerNum++) {
				BaseNeuralBlock *curLayer = _layers[layerNum];
				int prevLayerSize = curLayer->GetPreviousSize();
				int curLayerSize = curLayer->GetSize();
				float *curLayerWeights = curLayer->GetWeights();
				float *curLayerBias = curLayer->GetBias();
				float *learnFactors = _learnFactors[layerNum];
				float *packageDerivative = _packageDerivative[layerNum];
				float *derivativeAverages = _derivativeAverages[layerNum];
				float *oldDeltaWeights = _oldDeltaWeights[layerNum];
				float *learnFactorsForBias = _learnFactorsForBias[layerNum];
				float *packageDerivativeForBias = _packageDerivativeForBias[layerNum];
				float *oldDeltaWeightsForBias = _oldDeltaWeightsForBias[layerNum];
				float *derivativeAveragesForBias = _derivativeAveragesForBias[layerNum];
				float curLearnSpeed = _properties->BaseLearnSpeed*_properties->FactorStrategy->GetFactor(_epochNumber);

				parallel_for( blocked_range<size_t>(0, curLayerSize, BPAModifyWeightsGrainSize),
				[=](const blocked_range<size_t>& r)
				{
					for (int i = r.begin(); i < r.end(); i++) {
						for (int j = 0; j < prevLayerSize; j++) {
							int weightIndex = prevLayerSize*i + j;

							float lastDerivativeAverage = derivativeAverages[weightIndex];
							float partialDerivative = _packageFactor*packageDerivative[weightIndex] -
								_properties->Regularization->GetDerivative(curLayerWeights[weightIndex]);
							packageDerivative[weightIndex] = 0.0f;
							learnFactors[weightIndex] = (lastDerivativeAverage*partialDerivative > 0.0f) ?
								fminf(learnFactors[weightIndex] + _properties->SpeedBonus, _properties->SpeedUpBorder): 
								fmaxf(learnFactors[weightIndex]*_properties->SpeedPenalty, _properties->SpeedLowBorder);
							derivativeAverages[weightIndex] = _properties->AverageLearnFactor*partialDerivative + 
								(1.0f - _properties->AverageLearnFactor)*lastDerivativeAverage;

							float oldDelta = _properties->Momentum*oldDeltaWeights[weightIndex];
							float newDelta = curLearnSpeed*learnFactors[weightIndex]*partialDerivative + oldDelta;
							oldDeltaWeights[weightIndex] = newDelta;
							curLayerWeights[weightIndex] += (1.0f + _properties->Momentum)*newDelta;
						}

						float lastDerivativeAverageForBias = derivativeAveragesForBias[i];
						float partialDerivativeForBias = _packageFactor*packageDerivativeForBias[i];
						packageDerivativeForBias[i] = 0.0f;
						learnFactorsForBias[i] = (lastDerivativeAverageForBias*partialDerivativeForBias > 0.0f) ?
							fminf(learnFactorsForBias[i] + _properties->SpeedBonus, _properties->SpeedUpBorder):
							fmaxf(learnFactorsForBias[i]*_properties->SpeedPenalty, _properties->SpeedLowBorder);
						derivativeAveragesForBias[i] = _properties->AverageLearnFactor*partialDerivativeForBias + 
							(1.0f - _properties->AverageLearnFactor)*lastDerivativeAverageForBias;

						float oldDeltaForBias = _properties->Momentum*oldDeltaWeightsForBias[i];
						float newDeltaForBias = curLearnSpeed*learnFactorsForBias[i]*partialDerivativeForBias + oldDeltaForBias;
						oldDeltaWeightsForBias[i] = newDeltaForBias;
						curLayerBias[i] += (1.0f + _properties->Momentum)*newDeltaForBias;
					}
				});
			}
		}

		void BackPropagationAlgorithm::LocalGradientForOutputLayer(float *gradientsOutput, ActivationFunction *function, float *state, const float *errors,
				float *nextLayerGradients, float *nextLayerOldWeights, int curLayerSize, int nextLayerSize) {
			
			function->CalculateFirstDerivative(gradientsOutput, errors, state, curLayerSize);
		}

		void BackPropagationAlgorithm::LocalGradientForHiddenLayer(float *gradientsOutput, ActivationFunction *function, float *state, const float *errors,
				float *nextLayerGradients, float *nextLayerOldWeights, int curLayerSize, int nextLayerSize) {
			for (int neuronNum = 0; neuronNum < curLayerSize; neuronNum++) {
				gradientsOutput[neuronNum] = 0.0f;
			}

			parallel_for( blocked_range<size_t>(0, nextLayerSize, BPALocalGradientGrainSize),
			[=](const blocked_range<size_t>& r)
			{
				for (int j = r.begin(); j < r.end(); j++) {
					float nextLayerGradient = nextLayerGradients[j];
					for (int neuronNum = 0; neuronNum < curLayerSize; neuronNum++) {
						gradientsOutput[neuronNum] += nextLayerGradient*nextLayerOldWeights[curLayerSize*j + neuronNum];
					}
				}
			});

			function->CalculateFirstDerivative(gradientsOutput, state, curLayerSize);
		}
	}
}

File: NeuralNet/NeuralNetTypes/BlockType/NeuralNets/MultyLayerPerceptron/TrainMethods/BackPropagationAlgorithm.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using StandardTypes;

namespace NeuralNet.MultyLayerPerceptron {
	public sealed class BackPropagationAlgorithm : TrainMethod {
		private readonly RandomAccessIterator<TrainPair> _trainDataIterator;
        private readonly IList<TrainPair> _testData;
		private ITrainProperties _properties;
		private MultyLayerPerceptron _neuralNet;
		private BaseNeuralBlock[] _layers;
		private int[] _neuronNetLayersStruct;
		private float _packageFactor;
		private int _outputLayerNum;
		private float[] _neuronNetOutput;
		private float[] _neuronNetInput;
		private float[][] _oldDeltaWeights;
		private float[][] _derivativeAverages;
		private float[][] _packageDerivative;
		private float[][] _learnFactors;
		private float[][] _oldDeltaWeightsForBias;
		private float[][] _derivativeAveragesForBias;
		private float[][] _packageDerivativeForBias;
		private float[][] _learnFactorsForBias;
		private float[] _gradients;
		private float[] _gradientsIntermediate;
		private int _epochNumber;
		private int _packagesCount;

		public BackPropagationAlgorithm(IList<TrainPair> trainData) {
			_trainDataIterator = new RandomAccessIterator<TrainPair>(trainData);
		}

        public BackPropagationAlgorithm(IList<TrainPair> trainData, IList<TrainPair> testData) {
            _trainDataIterator = new RandomAccessIterator<TrainPair>(trainData);
            _testData = testData;
        }

		public override void InitilazeMethod(INeuralNet neuralNet, ITrainProperties trainProperties) {
			if (!(neuralNet is MultyLayerPerceptron)) {
				throw new ArgumentException("Neural net has other structure");
			}
			_neuralNet = (MultyLayerPerceptron) neuralNet;
			_properties = trainProperties;
			_layers = _neuralNet.Layers;
			_neuronNetLayersStruct = _neuralNet.GetLayersStruct();
			_outputLayerNum = _layers.Length - 1;
			_packageFactor = 1.0f/_properties.PackageSize;
			_packagesCount = CalculatePackagesCount();
			AllocateMemory();
			ProcessSate = IterativeProcessState.NotStarted;
		}

		public override ITrainProperties Properties {
			get { return _properties; }
		}

		private void AllocateMemory() {
			int maxGradientSize;
			FindMaxSize(out maxGradientSize);

			_gradients = new float[maxGradientSize];
			_gradientsIntermediate = new float[maxGradientSize];

			_neuronNetOutput = new float[_layers[_outputLayerNum].Size];

			_oldDeltaWeights = new float[_layers.Length][];
			_derivativeAverages = new float[_layers.Length][];
			_packageDerivative = new float[_layers.Length][];
			_learnFactors = new float[_layers.Length][];
			for (var i = 0; i < _layers.Length; i++) {
				var weightsCount = _layers[i].GetWeights()[0].Length;
				_oldDeltaWeights[i] = new float[weightsCount];
				_derivativeAverages[i] = new float[weightsCount];
				_packageDerivative[i] = new float[weightsCount];
				_learnFactors[i] = new float[weightsCount];
				for (var j = 0; j < weightsCount; j++) {
					_learnFactors[i][j] = 1f;
				}
			}

			_oldDeltaWeightsForBias = new float[_layers.Length][];
			_derivativeAveragesForBias = new float[_layers.Length][];
			_packageDerivativeForBias = new float[_layers.Length][];
			_learnFactorsForBias = new float[_layers.Length][];
			for (var i = 0; i < _layers.Length; i++) {
				var layerSize = _layers[i].Size;
				_oldDeltaWeightsForBias[i] = new float[layerSize];
				_derivativeAveragesForBias[i] = new float[layerSize];
				_packageDerivativeForBias[i] = new float[layerSize];
				_learnFactorsForBias[i] = new float[layerSize];
				for (var j = 0; j < layerSize; j++) {
					_learnFactorsForBias[i][j] = 1f;
				}
			}
		}

		private void FindMaxSize(out int maxGradientSize) {
			maxGradientSize = 0;
			for (var i = 1; i < _neuronNetLayersStruct.Length; i++) {
				var curLayerSize = _neuronNetLayersStruct[i];
				if (curLayerSize > maxGradientSize) {
					maxGradientSize = curLayerSize;
				}
			}
		}

		protected override void RunIterativeProcess() {
			var trainErrorValue = _properties.Epsilon + 1.0f;
			var testErrorValue = float.NaN;
			var minTestErrorValue = float.MaxValue;
			_epochNumber = 1;
			while ((ProcessSate == IterativeProcessState.InProgress) && 
				(trainErrorValue > _properties.Epsilon) && 
				(_epochNumber <= _properties.MaxIterationCount) && 
				(float.IsNaN(testErrorValue) || Math.Abs(testErrorValue - minTestErrorValue) < _properties.CvLimit)) {
				
				TrainEpoch();

				trainErrorValue = TestModel(_trainDataIterator.Collection);
			    testErrorValue = TestModel(_testData);
				if (!float.IsNaN(testErrorValue) && (testErrorValue < minTestErrorValue)) {
					minTestErrorValue = testErrorValue;
				}

                OnIterationCompleted(new IterationCompletedEventArgs(_epochNumber, trainErrorValue, testErrorValue));
				_epochNumber++;
			}
			OnIterativeProcessFinished(new IterativeProcessFinishedEventArgs(_epochNumber));
		}

		protected override void ApplyResults() {
			if (ProcessSate == IterativeProcessState.Finished) {
				ClearReference();
			}
		}

		private void ClearReference() {
			_properties = null;
			_neuralNet = null;
			_layers = null;
			_neuronNetLayersStruct = null;
			_neuronNetOutput = null;
			_neuronNetInput = null;
			_oldDeltaWeights = null;
			_gradients = null;
			_gradientsIntermediate = null;
			_derivativeAverages = null;
			_packageDerivative = null;
			_oldDeltaWeightsForBias = null;
			_derivativeAveragesForBias = null;
			_packageDerivativeForBias = null;
			_learnFactors = null;
			_learnFactorsForBias = null;
		}

		private int CalculatePackagesCount() {
			var packagesCount = _trainDataIterator.Size()/_properties.PackageSize;
			if (_trainDataIterator.Size()%_properties.PackageSize != 0) {
				packagesCount++;
			}
			return packagesCount;
		}

	    private float TestModel(IList<TrainPair> data) {
            if (data == null || data.Count == 0) {
	            return float.NaN;
	        }
	        else {
                var sumError = 0.0f;
                for (var i = 0; i < data.Count; i++) {
                    var testPair = data[i];
                    _neuronNetInput = testPair.Input;
                    _neuralNet.Predict(_neuronNetInput, _neuronNetOutput);
                    sumError += _properties.Metrics.Calculate(testPair.Output, _neuronNetOutput);
                }
                return sumError / data.Count;
	        }
	    }

		private void TrainEpoch() {
			_trainDataIterator.RefreshRandomAccess();
			for (var i = 0; i < _packagesCount; i++) {
				TrainPackage();
			}
		}

		private void TrainPackage() {
			for (var i = 0; i < _properties.PackageSize; i++) {
				var trainPair = _trainDataIterator.Next();
				_neuronNetInput = trainPair.Input;
				_neuralNet.Predict(_neuronNetInput, _neuronNetOutput);
				var outputPartialDerivaitves = _properties.Metrics.CalculatePartialDerivaitve(trainPair.Output, _neuronNetOutput);
				CollectWeightsDelta(outputPartialDerivaitves);
			}
			ModifyWeightsOfNeuronNet();
		}

		private void CollectWeightsDelta(float[] outputPartialDerivaitves) {
			const int firstLayerNumber = 0;
			var lastHiddenLayerNumber = _outputLayerNum - 1;

			CollectWeightsDeltaOfLayer(lastHiddenLayerNumber + 1, LocalGradientForOutputLayer, outputPartialDerivaitves);
			for (var layerNumber = lastHiddenLayerNumber; layerNumber >= firstLayerNumber; layerNumber--) {
				CollectWeightsDeltaOfLayer(layerNumber, LocalGradientForHiddenLayer, null);
			}
		}

		private void CollectWeightsDeltaOfLayer(int layerNum, LocalGradientDelegate localGradientfunction, float[] outputPartialDerivaitves) {
			var curLayer = _layers[layerNum];
			var prevLayerState = (layerNum > 0) ? _layers[layerNum - 1].GetState() : _neuronNetInput;
			var curGradients = _gradientsIntermediate;
			var nextGradients = _gradients;
			var nextLayerSize = (layerNum < _outputLayerNum) ? _neuronNetLayersStruct[layerNum + 2] : 0;
			var nextLayerWeights = (layerNum < _outputLayerNum) ? _layers[layerNum + 1].GetWeights()[0] : null;
			var packageDerivative = _packageDerivative[layerNum];
			var packageDerivativeForBias = _packageDerivativeForBias[layerNum];

			localGradientfunction(curGradients, curLayer.GetActivationFunction(), curLayer.GetState(), 
				outputPartialDerivaitves, nextGradients, nextLayerWeights, nextLayerSize);
			Parallel.For(0, curLayer.Size, i => {
				var localGradient = curGradients[i];
				for (var j = 0; j < prevLayerState.Length; j++) {
					packageDerivative[prevLayerState.Length*i + j] -= localGradient*prevLayerState[j];
				}
				packageDerivativeForBias[i] -= localGradient;
			});
			_gradients = curGradients;
			_gradientsIntermediate = nextGradients;
		}

		private void ModifyWeightsOfNeuronNet() {
			const int firstLayerNumber = 0;
			for (var layerNum = firstLayerNumber; layerNum <= _outputLayerNum; layerNum++) {
				var curLayer = _layers[layerNum];
				var curLayerWeights = curLayer.GetWeights()[0];
				var curLayerBias = curLayer.GetBias();
				var learnFactors = _learnFactors[layerNum];
				var packageDerivative = _packageDerivative[layerNum];
				var derivativeAverages = _derivativeAverages[layerNum];
				var oldDeltaWeights = _oldDeltaWeights[layerNum];
				var learnFactorsForBias = _learnFactorsForBias[layerNum];
				var packageDerivativeForBias = _packageDerivativeForBias[layerNum];
				var oldDeltaWeightsForBias = _oldDeltaWeightsForBias[layerNum];
				var derivativeAveragesForBias = _derivativeAveragesForBias[layerNum];
				var prevLayerSize = _neuronNetLayersStruct[layerNum];
				var curLearnSpeed = _properties.BaseLearnSpeed*_properties.LearnFactorStrategy.GetFactor(_epochNumber);

				Parallel.For(0, curLayer.Size, i => {
					for (var j = 0; j < prevLayerSize; j++) {
						var weightIndex = prevLayerSize*i + j;

						var lastDerivativeAverage = derivativeAverages[weightIndex];
						var partialDerivative = _packageFactor*packageDerivative[weightIndex] - 
							_properties.Regularization.GetDerivative(curLayerWeights[weightIndex]);
						packageDerivative[weightIndex] = 0.0f;
						learnFactors[weightIndex] = lastDerivativeAverage*partialDerivative > 0.0f ?
							Math.Min(learnFactors[weightIndex] + _properties.SpeedBonus, _properties.SpeedUpBorder) :
							Math.Max(learnFactors[weightIndex]*_properties.SpeedPenalty, _properties.SpeedLowBorder);
						derivativeAverages[weightIndex] = _properties.AverageLearnFactor*partialDerivative + 
							(1.0f - _properties.AverageLearnFactor)*lastDerivativeAverage;

						var oldDelta = _properties.Momentum*oldDeltaWeights[weightIndex];
						var newDelta = curLearnSpeed*learnFactors[weightIndex]*partialDerivative + oldDelta;
						oldDeltaWeights[weightIndex] = newDelta;
						curLayerWeights[weightIndex] += (1f + _properties.Momentum)*newDelta;
					}

					var lastDerivativeAverageForBias = derivativeAveragesForBias[i];
					var partialDerivativeForBias = _packageFactor*packageDerivativeForBias[i];
					packageDerivativeForBias[i] = 0.0f;
					learnFactorsForBias[i] = lastDerivativeAverageForBias*partialDerivativeForBias > 0.0f ? 
						Math.Min(learnFactorsForBias[i] + _properties.SpeedBonus, _properties.SpeedUpBorder) : 
						Math.Max(learnFactorsForBias[i]*_properties.SpeedPenalty, _properties.SpeedLowBorder);
					derivativeAveragesForBias[i] = _properties.AverageLearnFactor*partialDerivativeForBias + 
						(1.0f - _properties.AverageLearnFactor)*lastDerivativeAverageForBias;

					var oldDeltaForBias = _properties.Momentum*oldDeltaWeightsForBias[i];
					var newDeltaForBias = curLearnSpeed*partialDerivativeForBias + oldDeltaForBias;
					oldDeltaWeightsForBias[i] = newDeltaForBias;
					curLayerBias[i] += (1f + _properties.Momentum)*newDeltaForBias;
				});
			}
		}

		private static void LocalGradientForOutputLayer(float[] gradientsOutput, IActivationFunction function, float[] state, float[] errors,
			float[] nextLayerGradients, float[] nextLayerOldWeights, int nextLayerSize) {

			function.CalculateFirstDerivative(gradientsOutput, errors, state);
		}

		private static void LocalGradientForHiddenLayer(float[] gradientsOutput, IActivationFunction function, float[] state, float[] errors,
			float[] nextLayerGradients, float[] nextLayerOldWeights, int nextLayerSize) {
			var curLayerSize = state.Length;
			
			for (var neuronNum = 0; neuronNum < curLayerSize; neuronNum++) {
				gradientsOutput[neuronNum] = 0.0f;
			}

			for (var j = 0; j < nextLayerSize; j++) {
				var nextLayerGradient = nextLayerGradients[j];
				for (var neuronNum = 0; neuronNum < curLayerSize; neuronNum++) {
					gradientsOutput[neuronNum] += nextLayerGradient*nextLayerOldWeights[curLayerSize*j + neuronNum];
				}
			}

			function.CalculateFirstDerivative(gradientsOutput, state);
		}

		private delegate void LocalGradientDelegate(float[] gradientsOutput, IActivationFunction function, float[] state, float[] errors,
			float[] nextLayerGradients, float[] nextLayerOldWeights, int nextLayerSize);
	}
}